lib/linked_list.rb

```diff
@@ -18,128 +18,237 @@ def initialize
+      return nil if @head.nil?
+      current_node = @head
+      previous_node = nil
+      if current_node.data == value
+        @head = current_node.next
+      else
+        until current_node.data == value
+          previous_node = current_node
+          current_node = current_node.next
+        end
+        if current_node.next.nil?
+          previous_node.next = nil
+        else
+          previous_node.next = current_node.next
+        end
+      end
     end
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def reverse
-      raise NotImplementedError
+      current_node = @head
+      previous_node = nil
+      while current_node
+        next_node = current_node.next
+        current_node.next = previous_node
+        previous_node = current_node
+        current_node = next_node
+      end
+      @head = previous_node
     end
 
     # method that returns the value of the last node in the linked list
     # returns nil if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last
-      raise NotImplementedError
+      return nil if @head.nil?
+
+      list_length = self.length
+      return self.get_at_index(list_length - 1)
+
     end
-  
+
     ## Advanced Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_middle_value
-      raise NotImplementedError
+      # this method will return the value of the element at the index position
+      # represented by integer division of the list length by 2, regardless of
+      # whether the length is odd or even i.e. the "first middle" value will be
+      # returned for a list with even length.
+      return nil if self.length == 0
+
+      list_length = self.length
+      return self.get_at_index(list_length/2)
+
     end
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_nth_from_end(n)
-      raise NotImplementedError
+      return nil if n > self.length - 1
+      total_nodes = self.length - 1
+      nth_from_front = total_nodes - n
+      return get_at_index(nth_from_front)
     end
 
     # checks if the linked list has a cycle. A cycle exists if any node in the
     # linked list links to a node already visited.
     # returns true if a cycle is found, false otherwise.
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def has_cycle
-      raise NotImplementedError
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
+    def has_cycle?

```

Comment:
This only works if the cycle links back to the head, what if the cycle occurs further down.